### PR TITLE
Add RegexOptions.AnyNewLine via parser lowering

### PR DESCRIPTION
**Pull request by danmoseley**
## Motivation

.NET's `Regex` class hardcodes `\n` as the only newline character. With `RegexOptions.Multiline`, `$` matches before `\n` but not before `\r`, `\r\n`, or Unicode line breaks. This is ["by far one of the biggest gotchas"](https://github.com/dotnet/runtime/issues/25598) with `System.Text.RegularExpressions`:

```csharp
// BUG: on a file with Windows \r\n line endings, .+$ captures trailing \r
var match = Regex.Match("foo\r\nbar", ".*$", RegexOptions.Multiline);
// match.Value == "foo\r" — not "foo"!
```

Users are forced into fragile workarounds like `\r?$` or `(\r\n|\n)` to handle mixed line endings. Real-world NuGet packages show how common this is — from the [real-world regex patterns dataset](https://github.com/dotnet/runtime-assets/blob/main/src/System.Text.RegularExpressions.TestData/Regex_RealWorldPatterns.json):

- **`(\r\n|\n)`** (18,474 packages) — CSV parser manually matching both line endings
- **`\r?\n`** in PEM key parsing (1,964 packages) — `\r?\n` sprinkled throughout with `Multiline`, e.g., `^-+ *BEGIN (?<keyName>\w+) PRIVATE KEY *-+\r?\n...`
- **`$(\r?\n)?`** in assembly attribute matching (2,108 packages) — `(\s*\[\s*assembly:\s*(?:.*)\s*\]\s*$(\r?\n)?)` using `Multiline`
- **`[\r\n]+`** (2,422 packages) — matching any newline character

These workarounds are error-prone, don't compose well with `^` and `$` anchors, and miss Unicode newlines (`\u0085`, `\u2028`, `\u2029`).

## Summary

Implements `RegexOptions.AnyNewLine` ([api-approved](https://github.com/dotnet/runtime/issues/25598)) which makes `$`, `^`, `\Z`, and `.` recognize all Unicode line boundaries: `\r`, `\r\n`, `\n`, `\u0085` (NEL), `\u2028` (LS), `\u2029` (PS) — consistent with [Unicode TR18 RL1.6](https://unicode.org/reports/tr18/#RL1.6) and PCRE2's `(*ANY)` behavior.

With `AnyNewLine`, the example above just works:

```csharp
var match = Regex.Match("foo\r\nbar", ".*$", RegexOptions.Multiline | RegexOptions.AnyNewLine);
// match.Value == "foo" ✓
```

## Approach: Parser Lowering

All logic lives in `RegexParser.cs` — no changes to the interpreter, compiler, or source generator engines. Each affected construct is lowered into an equivalent `RegexNode` sub-tree:

| Construct | Lowered to |
|-----------|-----------|
| `$` (no Multiline) / `\Z` | `(?=\r\n\z\|\r?\z)\|(?<!\r)(?=\n\z)\|(?=[\u0085\u2028\u2029]\z)` |
| `$` (Multiline) | `(?=\r\n\|\r\|[\u0085\u2028\u2029]\|\z)\|(?<!\r)(?=\n)` |
| `^` (Multiline) | `(?<=\A\|\r\n\|\n\|[\u0085\u2028\u2029])\|(?<=\r)(?!\n)` |
| `.` | `[^\r\n\u0085\u2028\u2029]` (but `Singleline` takes precedence) |

Key design choices:
- **`\r\n` is atomic**: `$` never matches between `\r` and `\n`. This is enforced with lookbehind/lookahead guards.
- **`Singleline` takes precedence**: `.` with `Singleline | AnyNewLine` matches everything (including newlines), consistent with `Singleline`'s documented behavior.
- **`\A` and `\z` are unaffected**: absolute start/end anchors don't change.
- **Incompatible with `NonBacktracking` and `ECMAScript`**: throws `ArgumentOutOfRangeException` (lowered patterns use lookaround).

### Production code
- `RegexOptions.cs` — add `AnyNewLine = 0x0800`
- `RegexParser.cs` — lowering methods `AnyNewLineEndZNode()`, `AnyNewLineEolNode()`, `AnyNewLineBolNode()`, plus `.` handling
- `RegexCharClass.cs` — add `NotNewLineOrCarriageReturnClass` constant
- `Regex.cs` / `RegexCompilationInfo.cs` — validation

### Tests
- ~120 new test cases covering dot, anchors (`$`, `^`, `\Z`), `RightToLeft`, `Singleline`, `Multiline`, `Replace`, `Split`, `Count`, `EnumerateMatches`, `NonBacktracking` rejection, edge cases (adjacent newlines, empty lines, all-newline strings), and PCRE2-inspired scenarios

Fixes dotnet/runtime#25598

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs [hunk 1]
```diff
@@ -3232,4 +3232,752 @@ public static IEnumerable<object[]> BalancingGroup_Various_MemberData()
...
+            // Replace each line with its line number using a MatchEvaluator
+            Regex numberer = await RegexHelpers.GetRegexAsync(engine, @"^.+$", MA);
+            int lineNum = 0;
+            string numbered = numberer.Replace(input, _ => (++lineNum).ToString());
+            Assert.Equal("1\r\n2\r3\n4", numbered);
+
+            // Split on multiline $ with AnyNewLine — each line ending is a split point
+            Regex splitter = await RegexHelpers.GetRegexAsync(engine, @"$\r?\n?", MA);
+            string[] parts = splitter.Split("ab\rcd\r\nef");
+            Assert.Equal(new[] { "ab", "cd", "ef", "" }, parts);
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs]: without anynewline, would split give an empty final match like it's doing here?
